### DISCUSSION TITLE
Convert Text to Numbers/Recognize Numbers

### DISCUSSION THREAD

**Discussion by thoughtcastapp** [speechly/speechly#250] [Q&A]
Hi, I am using Speechly as an alternative to the SpeechKit built into iOS...I am trying to convert a number to digits, so if the user says "five thousand two hundred and nineteen" it shows 5219, and not the actual words. Did some searching, but couldn't find a solution (and a good solution for this doesn't seem to exist otherwise)...anything I can do potentially to get this working?

**Comment by bigdatabaracus** [comment]
Thanks @thoughtcastapp for the questions. Are you building a voice UI / assistant or are you using Speechly to do transcription? In other words, what's your use case? The current way of getting digits out of Speechly is using the [number](https://docs.speechly.com/reference/entity-data-types#number) entity type. For open ended transcription that's however rather cumbersome to get working. Good for you to know however, that due to popular demand we are working right to get a more straight forward  support for digits also for the transcription use case.

**Comment by thoughtcastapp** [comment]
@bigdatabaracus it is primarily a maths app, we were using the built in speech recognizer (speech kit) but it doesn't work on one of our targets so we had to switch...what speech kit was doing well is any number longer than a single digit, it was automatically changing it to actual numbers (so we could process these as Int/Double/etc), but speechly isn't doing that for us automatically....

**Comment by cfry** [comment]
I am hoping to get numbers returned as entities in the segment data structure passed to
client.onSegmentChange in my browser client JS code.
I don't want to declare some intent, just give me a list of all numbers in every utterance.

I see indications of lots of functionality scattered around Speechly doc,
such as https://docs.speechly.com/reference/entity-data-types/#number
but what I don't see is end-to-end examples.
- Is the code snippet supposed to be for my Javascript code?
- Some browser-client initialization in VAD?
- some dashboard configuration such as in my profile for training data?
- some SAL language code ... but where does that go?
- https://docs.speechly.com/reference/standard-variables has an impressive array of entities that can be recognize
and you document the variables that are associated with these entities. But where, specifically, do
I put these variables?
- On my app  dashboard page, under the training data tab, in the left hand column, in the Entity section,
I have selected "number" from the dropdown and typed in an entity name, and
clicked the "deploy" button, but the segments passed to the callback client.onSegmentChange
have an empty array for entities. 
Iv'e left the type-inn area under "Text adaptation" empty.
Its "show example utterances" link is useless.

Overall the doc has skimpy fragmented pieces that are disconnected from each other.
Please make your documentation useful by connecting them.

I have been able to get basic speech reco working and the actual speech to text is quite good.

**Reply by teelisyys** [reply]
Hi @cfry !

Thank you for trying out Speechly and thank you for your questions!

For full, working end-to-end examples please check out examples and demos in our [Speechly monorepo ](https://github.com/speechly/speechly). For example the the [browser-client example](https://github.com/speechly/speechly/blob/main/examples/browser-client-example/src/index.ts#L32) contains setup for the VAD.

* * * 

Our entity and intent detection is based on example training utterances that you provide. The utterances are written in SAL language to the the "Training Data" / "Text adaptation" text area. If no utterances are provided, then no entity or intent detection will happen. The type of the Entity defines what kind of post processing is done to the recognised entity value, e.g. with _number_ the words are transformed to a number written with digits (eg `twenty two → 22`)

The predefined variables can be used in the SAL example utterances, just like any variables that you define yourself.

For example with the following SAL in the "Training Data" / "Text adaptation" and the _count_ entity set to _number_ type

~~~
*statement I have $SPEECHLY.SMALL_CARDINAL_NUMBER(count) [lemons|mandarins|limes]
~~~

an utterance like _i have seventy two lemons_ would produce an entity with the count as 72:

~~~
 "entities": [
    {
      "type": "count",
      "value": "72",
      "startPosition": 2,
      "endPosition": 4,
      "isFinal": true
    }
  ]
~~~

The "Show sample utterances" button provides a sample of utterances that the written SAL training data would produce. As the SAL can get complex, this is useful to check if there are any unwanted training utterances.

Please note that the example utterances should closely reflect the actual usage in your specific use case.

**Comment by cfry** [comment]
Thanks for the quick response.
I got number entity reco working!
(despite bad UI, bad lang design, and bad doc.)

Almost everyone does bad UI and bad lang design,
so you've got a lot of company.
The world wide web standard for doc on API's is "none".
At least you have *some* doc, but its hard to write good
doc for bad UI and bad lang design.

Much better is just make those things better, then doc becomes less
necessary
(but rarely unnecessary with a complex, high-functionality product like
yours)

Anyway, I want to help you make your product easier to use so the world
can take advantage of your highly accurate speech reco.

First, UI:
on the page: https://api.speechly.com/dashboard
under the tab: "Training Data"
This isn't "training data" in the normal sense of today's AI use of the
term.
A better name would be "entity specification" or something like that.
When you said "training data" I got scared: I want number reco,
so I'd have to enter at least one example of every number I wanted
recognized
(if this was normal "training data".)???
Don't use the term "training data". It will scare your users, as it did me.

Next: the left hand column forces me to declare all the intents I want to
use
and all the entities. But this is completely redundant with the
text under "Text Adaptation". Just get rid of the left hand column
and pull the data out of the "Text Adaptation" pane.
The left column is just confusing.

"Text Adaptation" column header above the big text area.
This is the ONLY place on your website that I've seen "Text Adaptation".
Its meaningless. Better would be
SAL Intents and Entities (or something like that.) See below on "SAL
Configuration"

Next up, Language Design.
Here's what worked for me in the Text Adaption type in:
*relative_move down $SPEECHLY.NUMBER(num)

Using cryptic symbols like * and $ is bad lang design.
Sure they're concise, but users don't care about conciseness as much as
understandability. The experts at Speechly may type in SAL all day long
and want to save a few chars, but your users don't.
Your users have to know lots of other languages
In my case, JavaScript, HTML, CSS, JSON, URL and a bunch more.
We don't need another cryptographic language to learn.
Replace "*" with "intent:"
Replace "$" with "type:"
   You call "$SPEECHLY.NUMBER" a "Standard Variable".  Wrong. Its a type,
   its not a variable.
   Making it worse, you've got a syntax:
   $SPEECHLY.NUMBER(num)
   where you're passing a variable an argument of "num" but the syntax
   looks like a method call. Its NOT a method call (according to you)
    so don't make it look like one.
   Its a type. But rather than (num) being some kind of type modifier
   (ie a number that's "positive"), its really a label for the recognized
number.
   Don't use function call syntax for a type.
   Now that "num" is really what I'd call a label for the recognized number.
  OK so call it a "label:"
Putting the above suggestions together for my example we get:
intent:relative_move down type:SPEECHLY.NUMBER label:num
See? You can actually read it and it makes some sense.

UI bug:
I wanted to get some examples for your cryptographic SAL.
Fortunately in the upper left of the page you have
"show sample utterances"
when the Text Adaptation pane is empty it essentially gives a useless
error message.
Now that I have some text in it,
it gives some content that looks like it might be useful (good)
but the header for the window that pops up is:
"Utterances sampled from the SAL configuration"
Huh? SAL configuration? You mean Text Adaptation?
I suspect you're using two different terms for the same thing.
Always wrong. Synonyms are evil.  Never use them in doc
or UI unless you're defining a term.
I like "SAL Configuration" much better than "Text Adaptation"
(You eliminate synonyms by having a company GLOSSARY
  of terms. Having a glossary is the easiest way to improve your product.
  Publish it on your web page!)

I guess you are saying with your header that each line in the window
is an utterance. Well if those are utterances, what would you call
just plain spoken text without the square brackets and parens?
(beats me).
So I'd say those lines aren't utterances". Maybe they are
"SAL patterns" or something.
Better would be, just have the plain text for each line in that pop -up
window

"sampled" is wrong. A sample is one of a set of things.
Those aren't samples, more like patterns.
And they aren't "sampled", they're generated from the SAL config.
So title the window something like:
"Patterns generated from your SAL configuration"

As for when you have no text under Text Adaptation,
put in some useful examples of text you COULD have.
That's what I was expecting.

I could spend a long time trying to figure out your almost useless doc.
But its not worth documenting such lousy UI and lang design.
Get UI and lang design right, THEN work on the doc.

I will help you more if you're willing to try and make your
underlying great software actually useful to many more people.
Its comparatively easy to make the changes I propose than
tweaking your core speech-reco algorithm,
and you'll get a lot more use of your product.

**Reply by teelisyys** [reply]
Hi @cfry ! Thank you for your detailed feedback!

We strive constantly to improve our documentation, but as you point out, being familiar with our own system makes it harder for us to recognise issues an outsider finds obvious. We will go through your feedback carefully and try to address points you raised in the near future.

However, there are a few points I'd like to clarify, as I do not think they correctly reflect our system.

To us the left column is not redundant in the _training data_ tab. This is the only place where the entity data types can be defined. As the Speechly Annotation Language (SAL) has no concept of typing, it is a mere text template processor with annotation syntax. The _standard variables_ have no inherent typing information, in fact they are internally defined using the same SAL variable syntax available to end users. This is because for example date expressions have various possible post-processing conventions and there is no one-to-one mapping between the variables and types.

The data from the _training data_ tab is used to train our machine learning system. We also recommend using real-word example utterances whenever possible. The templating features are for cases when such data is not available, or only such a small amount of real-word data is available that it needs to be varied manually. The current subtitle (“Text adaptation”) is indeed somewhat misleading, as this data is used both for the SLU training and for [Text Adaptation](https://docs.speechly.com/basics/model-adaptation/#text-adaptation) of the ASR.

As for our design of our SAL language, please note that it is not a general purpose programming language, but a domain specific annotation language (similar to e.g. Markdown) with some template processing support built in. Its design influences and constraints thus differ from a general purpose programming language.

Of course we should communicate these points more clearly in our documentation.

**Comment by cfry** [comment]
On Tue, Jan 3, 2023 at 8:24 AM Juho Kilpikoski ***@***.***>
wrote:

> Hi @cfry <https://github.com/cfry> ! Thank you for your detailed feedback!
>
> We strive constantly to improve our documentation, but as you point out,
> being familiar with our own system makes it harder for us to recognise
> issues an outsider finds obvious. We will go through your feedback
> carefully and try to address points you raised in the near future.
>
thank you, That's the best I can hope for.

> However, there are a few points I'd like to clarify, as I do not think
> they correctly reflect our system.
>
> To us the left column is not redundant in the *training data* tab. This
> is the only place where the entity data types can be defined.
>

From https://docs.speechly.com/reference/sal  I quote:
"Entities are defined by [entity value](entity name) notation.

For example:

*show_products show [jeans](category)
And since the SAL is meant to go under Text Adaptation (though I can't find
any doc that
says that), then you can define entities in the right column of the
training data page
according to your doc. In fact, according to your doc, you CAN'T define an
entity in the left column because that isn't SAL.

In my example: *relative_move down $SPEECHLY.NUMBER(num)
it seems to me the entity has all the info that's in the left pane, ie the
entity's type and its label.
What other information is there in the left column that I'm missing?

You didn't mention the intent "declarations" in the left column but seems to
me all that's needed is to say *my_intent
and we're done "declaring" our intent.

As the Speechly Annotation Language (SAL) has no concept of typing,
>


Is  $SPEECHLY.NUMBER  part of SAL?
If not then I'm even more confused as you're mixing different languages on
the
same line of text.  Number is a type. Ask any programmer that might be
using your system.
*You* can redefine the word "type", but that doesn't help your users.

> it is a mere text template processor with annotation syntax. The *standard
> variables* have no inherent typing information,
>
This is like saying
let foo
has no inherent value.
But once you do
let foo = 2
then foo MEANS 2 where ever you use "foo"
I don't know what you mean by inherent, I guess.

> in fact they are internally defined using the same SAL variable syntax
> available to end users. This is because for example date expressions have
> various possible post-processing conventions and there is no one-to-one
> mapping between the variables and types.
>
Sure, just like there's no one to one mapping between any variable and any
value.
we can use let bar = 2  and we've broken the one-to-one mapping. OK by me.
foo still has a value of 2, and 2 is a number, so foo holds data that is a
number.

In normal code, you can have a snippet of code that creates a value and
copy & paste it
everywhere you want that, OR you can set a var to it and use the var.
If you have complex types, its fine to have a variable represent a type,
but its still semantically a type.

I like parameterized types. Its good that you have variables to hold the
common ones.
I should be able to use the variable as a type so as far as I'm concerned,
it IS the type.



> The data from the *training data* tab is used to train our machine
> learning system.
>
OK so there's an additional functionality to that tab that "training data"
captures,
but misses the one you document most as far as I can see, which is
describing the patterns.

> We also recommend using real-word example utterances whenever possible.
>
Real world utterances don't have square brackets in them.
Note from my previous message the heading to the pop up window of:
"Utterances sampled from the SAL configuration"
So the heading is saying that the below lines are Utterances and they
have square brackets in them, but, unless we have a different definition of
"real world",
real world utterance don't have square brackets in them, so your def of
Utterance in your email is different than your website.
If you aren't consistent with definitions, you're going to confuse a lot
more people than me.

> The templating features are for cases when such data is not available, or
> only such a small amount of real-word data is available that it needs to be
> varied manually.
>
Well ok, but at least for my purposes the "templating" are the only thing I
want to use.
Maybe I'm an unusual user in this regard but I'd guess not since I just
spent
20 minutes looking for an example of this "training data" that isn't
a "template" under https://docs.speechly.com/reference/ and a couple levels
deep
and couldn't find any.

I did find on https://docs.speechly.com/reference/sal,
"A SAL expression is an example utterance if it does not contain any
template notation. That is, it can only be expanded to itself.

All the following SAL expressions are example utterances:

*search show [red](color) [pants](product)
"
OK so "*search show [red](color) [pants](product)" does not contain any
template notation,
but then on the same page under "Template Notation" you have:
"A SAL expression is a template if it contains template notation and can be
expanded to an example utterance.

All of these SAL expressions are templates:

*search show {[red | green | blue](color)} [pants | shirts | shoes](product)
*book book a flight ![from $city(depart_city) | to $city(arrival_city)]
*greeting $all_greeting_phrases
"
so *search show [red](color) [pants](product)  is NOT a template but

*search show {[red | green | blue](color)} [pants | shirts |
shoes](product)   IS a template.
If it doesn't have curly braces it isn't a template?


The current subtitle (“Text adaptation”) is indeed somewhat misleading, as
> this data is used both for the SLU training and for Text Adaptation
> <https://docs.speechly.com/basics/model-adaptation/#text-adaptation> of
> the ASR.
>
OK thanks for that link. I missed it because it isn't under Reference.
What's unclear to me is, if a user speaks a sentence that ISN'T in the Text
adaptation,
it still gets recognized (I tried it) so why provide it?
Is it the case that if your speech reco is trying to decide which of two
sentences the
user said and one of them happens to be in the Text Adaptation, then
it uses that one? (If so, please Say so on the Text Adaptation page)
Is Text Adaptation the same as "Training Data"?
You need a glossary as I said in my previous message.



> As for our design of our SAL language, please note that it is not a
> general purpose programming language, but a domain specific annotation
> language (similar to e.g. Markdown) with some template processing support
> built in. Its design influences and constraints thus differ from a general
> purpose programming language.
>
yes certainly. It doesn't have to be as complex as a general purpose lang,
but I caution you,
that the vast majority of computer langs are not general purpose, though as
they
mature and get popular, they get more and more general purpose, but must
maintain
backwards compatibility (for users wanting it) and then
the language becomes "incrementally designed" which is to say "not designed"
which is to say "way unnecessarily difficult to use".

Obviously I'm not an expert in your system but here's
three other big ideas for you.

1. Instead of SAL syntax, use JSON.
Yes this will be more verbose, but will leave room for lots of new features
in a consistent
fashion. Also you won't have to document the syntax.
Also many tools can already
parse and dissect and even write JSON.
So for my one line of SAL in my app:
*relative_move down $SPEECHLY.NUMBER(num)
I suggested a more easily understood syntax of:
intent:relative_move down type:SPEECHLY.NUMBER label:num
OK if we convert that into JSON we get:
[{intent: "relative_move"},  "down", {type: "number", label: "num"}]
then we can do more like:
[{intent: "relative_move"},
  {text: ["down", "go down", "descend"]    //for alternatives
  {type: "number", label: "num",               //an entity
   min:0,                                                  //its "arity"
   max: 1}]                                               //its "arity"
for indicating that its optional or
min:1 max:1 (the default, required, or
min:2 max: 5  etc.

Yes this is more cumbersome but more standard, and one less
idiosyncratic language for your users to learn.
JSON isn't great, but it will be either less learning for your users
or learning that will be more easily applied to other things.
Actually they have to know it anyway to decode a segment
so not really something extra to learn.

I did read a bit more about SALs variables.
I see why you use the $. Like BASH, about the worst of popular prog langs.
You could still use the $ syntax in a string or better:
var:foo


2. Don't have client data entered on the server.
Let programmers write and keep the "source" for their data
within their program and SEND it to your server
in an "initialization" call from their code.
This lets all of the user's "program" to exist in one file
or at least their machine.
The user can back it up, and back up different versions.
It avoids at least some of the UI problems I articulated in previous email.
It makes sharing example code possible because its all in
one file that you can just use as is.
And in case you switch to JSON, then the user's program can
easily manipulate that data, programmatically producing
the templates and/or training data.
It will be a whole lot less confusing because as it is now, the
relationship between the "on server web page" data
and "my code" is not well documented but wouldn't
need to be nearly so much if it were all in one coherent file.

3. Just automatically do entity recognition and supply them to the user
by default.
Sure there might be some special entities that the user could declare,
 but if all your reco returned to the user was
["the whole sentence said", 123, 45, "2023/01/28 3:45"]
ie the entities are just in the order that they are in the utterance,
that would be really easy to learn and handle.
Maybe the user requests returning
locations, person_names, and some others you've got.
Its not hard for a programmer to detect a type from those values,
and we often don't need to label an entity if we just get the order.
The more complex info you return is useful, nice to have available,
but often doesn't matter. My case is one example.

> Of course we should communicate these points more clearly in our
> documentation.
>
This is going to be hard. Please work on the UI and lang first.
Plus programmers hate writing doc,
documentors don't understand the semantics enough to write good doc,
and auto-generated doc is often of negative value.
I write an awful lot of doc, but the better the UI and lang design I do,
the less doc I need to write.
Users (as far as I can tell) really don't like reading doc either.

I appreciate your response. I know a lot more about your system now.
